substratest/client.py

```diff
@@ -16,47 +16,67 @@ class _State:
     """
 
     def __init__(self):
-        self.datasets = []
-        self.test_data_samples = []
-        self.train_data_samples = []
-        self.objectives = []
-        self.algos = []
         self.aggregate_algos = []
-        self.composite_algos = []
-        self.traintuples = []
         self.aggregatetuples = []
+        self.algos = []
+        self.composite_algos = []
         self.composite_traintuples = []
-        self.testtuples = []
         self.compute_plans = []
+        self.datasets = []
+        self.objectives = []
+        self.test_data_samples = []
+        self.testtuples = []
+        self.train_data_samples = []
+        self.traintuples = []
 
     def _update_assets(self, asset_list_name, asset):
+        """Replace or insert asset in self.[asset_list_name]"""
         asset_list = getattr(self, asset_list_name)
-        setattr(self, asset_list_name, [
-            asset if asset.key == a.key else a
-            for a in asset_list
-        ])
+        new_asset_list = []
+        found = False
+        for a in asset_list:
+            if asset.key == a.key:
+                found = True
+                new_asset_list.append(asset)
+            else:
+                new_asset_list.append(a)
+        if not found:
+            new_asset_list.append(asset)
+        setattr(self, asset_list_name, new_asset_list)
```

Comment:
We could also use a dict to update the asset:

```python
existing_assets = {a.key: a for a in getattr(self, asset_list_name)}
existing_assets[asset.key] = asset
setattr(self, asset_list_name, list(existing_assets.values()))
```

Ideally, we should represent the assets using a dict (for performance reaon), but this could be done later.